timely_beliefs/beliefs/classes.py

```diff
@@ -276,18 +277,29 @@ def add_to_session(
         """
         # Belief timing is stored as the belief horizon rather than as the belief time
         belief_records = (
-            beliefs_data_frame.convert_index_from_belief_time_to_horizon()
-            .reset_index()
-            .to_dict("records")
+            beliefs_data_frame.convert_index_from_belief_time_to_horizon().reset_index()
         )
-        beliefs = [cls(sensor=beliefs_data_frame.sensor, **d) for d in belief_records]
+        beliefs = [
+            cls(sensor=beliefs_data_frame.sensor, **d)
+            for d in belief_records.to_dict("records")
+        ]
+
         if expunge_session:
             session.expunge_all()
-        if not allow_overwrite:
-            if bulk_save_objects:
-                session.bulk_save_objects(beliefs)
-            else:
-                session.add_all(beliefs)
+
+        if bulk_save_objects:
+            # serialize source and sensor
+            belief_records["source_id"] = belief_records["source"].apply(lambda x: x.id)
+            belief_records["sensor_id"] = belief_records.sensor.id
+            belief_records = belief_records.drop(columns=["source"])
+
+            session.execute(
+                insert(cls)
+                .values(belief_records.to_dict("records"))
+                .on_conflict_do_nothing()
```

Comment:
`if not allow_overwrite` then we should have conflicts resulting in failure, though.